### PR TITLE
Templates: Use theme mods rather than wp_theme tax

### PR DESCRIPTION
**Pull request by ockham**
Based on @carlomanf's https://github.com/WordPress/gutenberg/pull/31971. 

More details to follow.

Trac ticket: https://core.trac.wordpress.org/ticket/53354

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/block-template-utils.php [hunk 1]
```diff
@@ -17,24 +17,29 @@
  * @return WP_Block_Template|WP_Error Template.
  */
 function _build_template_result_from_post( $post ) {
-	$terms = get_the_terms( $post, 'wp_theme' );
+	$template_type = $post->post_type;
 
-	if ( is_wp_error( $terms ) ) {
-		return $terms;
+	if ( 'wp_template' !== $template_type ) {
+		return new WP_Error( 'template_wrong_post_type', __( 'An invalid post was provided for this template.', 'gutenberg' ) );
```
Comment (youknowriad) [inline src/wp-includes/block-template-utils.php:23]: We should remove the "Gutenberg" domain from the translations.

File: src/wp-includes/theme-templates.php [hunk 2]
```diff
@@ -1,69 +1,59 @@
...
- * @return string The original, desired slug.
+ * @param string $slug          Template slug.
+ * @param string $template_type wp_template.
+ *
+ * @return bool Whether the template is customized for the currently active theme.
  */
-function wp_filter_wp_template_unique_post_slug( $override_slug, $slug, $post_ID, $post_status, $post_type ) {
-	if ( 'wp_template' !== $post_type ) {
-		return $override_slug;
+function template_is_customized( $slug, $template_type = 'wp_template' ) {
```
Comment (youknowriad) [inline src/wp-includes/theme-templates.php:14]: do you think `block_template_is_customize` is a better name?

File: src/wp-includes/block-template-utils.php (hunk 1 above)
Comment (ockham) [reply src/wp-includes/block-template-utils.php:23]: We should write a static check and integrate it with CI to point these out 🤔

File: src/wp-includes/theme-templates.php (hunk 2 above)
Comment (ockham) [reply src/wp-includes/theme-templates.php:14]: Maybe, yeah. I was wondering if `*_is_customized` fully captured what this was doing anyway -- maybe `customized_block_template_exists` is more descriptive.

**Comment by ockham** [comment]
A little update for everyone following this: @youknowriad and I decided not to include this in the Beta 1, since I barely just got it in shape (and tests to pass). We'll allow some more time for testing and discussion of the potential impact.

It seems like the other discussion over at https://github.com/WordPress/gutenberg/pull/31971 got a bit heated. That's understandable, since it's a significant change to the inner workings of the templating system that (for a number of reasons) comes fairly late into the WP 5.8 release cycle.

I find both sides of the argument understandable: On one hand, it's desirable to avoid the requirement for a later migration (if we keep using the taxonomy, and in a later WP version decide to switch to theme mods). OTOH, this hasn't gone through the same amount of testing by Full-Site Editing folks as the taxonomy-based system, and we might lose "free" block template exports.

We might disagree on the impact and importance of these aspects, but let's keep in mind that ultimately, we all mean well for the project; we might have different opinions, and assign different weight to various aspects, but I'm sure we'll be able to arrive at a good conclusion if we respect each others' opinions ❤️

**Comment by ockham** [comment]
We decided to hold off on this change for the 5.8 release. See https://github.com/WordPress/gutenberg/pull/31971#issuecomment-859556344 for more details.

**Pull request closed without merging**